File: CarritoBlueJ/CarritodeCompras/Carrito.java
import lang.stride.*;
import java.util.List;
import java.util.ArrayList;

public class Carrito
{
    private static final double PORCENTAJE_IVA = 0.12;
    private Cliente cliente;
    private List<DetalleCarrito> items;

    public Carrito(Cliente cliente)
    {
        this.cliente = cliente;
        this.items =  new  ArrayList < DetalleCarrito > ();
    }

    public Cliente getCliente()
    {
        return cliente;
    }

    public List<DetalleCarrito> getItems()
    {
        return items;
    }

    public boolean agregarItem(Producto producto, int cantidad)
    {
        if (producto == null || cantidad <= 0) {
            return false;
        }
        if ( ! producto.reducirStock(cantidad)) {
            Notificacion aviso =  new  Notificacion("Administrador", "Sin stock suficiente de: " + producto.getNombre(), "hoy");
            aviso.enviar();
            return false;
        }
        for (final DetalleCarrito detalle : items) {
            if (detalle.getProducto().getId() == producto.getId()) {
                detalle.setCantidad(detalle.getCantidad() + cantidad);
                return true;
            }
        }
        items.add( new  DetalleCarrito(producto, cantidad));
        return true;
    }

    public boolean eliminarItem(int idProducto)
    {
        int i = 0;
        while (i < items.size()) {
            DetalleCarrito detalle = items.get(i);
            if (detalle.getProducto().getId() == idProducto) {
                detalle.getProducto().devolverStock(detalle.getCantidad());
                items.remove(i);
                return true;
            }
            i = i + 1;
        }
        return false;
    }

    public boolean estaVacio()
    {
        return items.isEmpty();
    }

    public double calcularSubtotal()
    {
        double subtotal = 0;
        for (final DetalleCarrito detalle : items) {
            subtotal = subtotal + detalle.calcularSubtotal();
        }
        return subtotal;
    }

    public double calcularImpuestos()
    {
        return calcularSubtotal() * PORCENTAJE_IVA;
    }

    public double calcularTotal()
    {
        return calcularSubtotal() + calcularImpuestos();
    }

    public void vaciarCarrito()
    {
        items.clear();
    }

    public boolean finalizarCompra(Pago pago)
    {
        if (estaVacio()) {
            return false;
        }
        boolean aprobado = pago.procesarPago(calcularTotal());
        if (aprobado) {
            vaciarCarrito();
        }
        return aprobado;
    }
}

File: CarritodeCompras/logica/Carrito.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CarritodeCompras.logica
{
    public class Carrito
    {
        private const double PORCENTAJE_IVA = 0.12;

        private Cliente cliente;
        private List<DetalleCarrito> items;

        public Carrito(Cliente cliente)
        {
            this.cliente = cliente;
            items = new List<DetalleCarrito>();
        }

        public Cliente Cliente { get { return cliente; } }
        public List<DetalleCarrito> Items { get { return items; } }
        public bool AgregarItem(Producto producto, int cantidad)
        {
            if (producto == null || cantidad <= 0)
            {
                return false;
            }
            if (!producto.ReducirStock(cantidad))
            {
                var avisoAdmin = new Notificacion("Administrador", $"Sin stock suficiente de: {producto.Nombre}");
                avisoAdmin.Enviar();
                var avisoCliente = new Notificacion("Cliente", $"El producto '{producto.Nombre}' no tiene stock disponible.");
                avisoCliente.Enviar();
                return false;
            }

            var existente = items.FirstOrDefault(d => d.Producto.Id == producto.Id);
            if (existente != null)
            {
                items.Remove(existente);
                items.Add(new DetalleCarrito(producto, existente.Cantidad + cantidad));
            }
            else
            {
                items.Add(new DetalleCarrito(producto, cantidad));
            }
            return true;
        }

        public bool EliminarItem(int idProducto)
        {
            var item = items.FirstOrDefault(d => d.Producto.Id == idProducto);
            if (item == null)
            {
                return false;
            }
            item.Producto.DevolverStock(item.Cantidad);
            items.Remove(item);
            return true;
        }

        public double CalcularSubtotal()
        {
            return items.Sum(d => d.CalcularSubtotal());
        }

        public double CalcularImpuestos()
        {
            return CalcularSubtotal() * PORCENTAJE_IVA;
        }

        public double CalcularTotal()
        {
            return CalcularSubtotal() + CalcularImpuestos();
        }

        public bool EstaVacio()
        {
            return items.Count == 0;
        }

        public void VaciarCarrito()
        {
            items.Clear();
        }

        public bool FinalizarCompra(Pago pago)
        {
            if (EstaVacio() || pago == null)
            {
                return false;
            }
            bool aprobado = pago.ProcesarPago(CalcularTotal());
            if (aprobado)
            {
                VaciarCarrito();
            }
            return aprobado;
        }
    }
}
